File: ios/WiretalkSDK/Sources/WiretalkSDK/WiretalkBridge.swift
import Foundation

public struct WiretalkBridgeEvent {
    public let type: String
    public let raw: [String: Any]

    public var visitorUid: String? { raw["visitor_uid"] as? String }
    public var conversationId: Int? { raw["conversation_id"] as? Int }
    public var unreadCount: Int { raw["count"] as? Int ?? 0 }
    public var messageId: Int? { raw["message_id"] as? Int }
    public var messageBody: String? { raw["body"] as? String }
    public var isOpen: Bool { raw["open"] as? Bool ?? false }
    public var tab: String? { raw["tab"] as? String }
}

public enum WiretalkBridge {
    public static func parse(_ payload: Any) -> WiretalkBridgeEvent? {
        let dict: [String: Any]?
        if let string = payload as? String,
           let data = string.data(using: .utf8),
           let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
            dict = json
        } else if let json = payload as? [String: Any] {
            dict = json
        } else {
            dict = nil
        }

        guard let dict, let type = dict["type"] as? String, !type.isEmpty else {
            return nil
        }

        return WiretalkBridgeEvent(type: type, raw: dict)
    }

    public static func command(type: String, fields: [String: Any] = [:]) -> String {
        var payload: [String: Any] = ["wiretalk": true, "type": type]
        fields.forEach { payload[$0.key] = $0.value }

        guard let data = try? JSONSerialization.data(withJSONObject: payload),
              let json = String(data: data, encoding: .utf8) else {
            return ""
        }

        return "window.dispatchEvent(new MessageEvent('message', { data: \(json) }));"
    }

    public static func open(tab: String = "chat") -> String {
        command(type: "widget:open", fields: ["tab": tab])
    }

    public static func close() -> String {
        command(type: "widget:close")
    }

    public static func toggle() -> String {
        command(type: "widget:toggle")
    }

    public static func updateVisitor(name: String? = nil, email: String? = nil, phone: String? = nil) -> String {
        var fields: [String: Any] = [:]
        if let name { fields["name"] = name }
        if let email { fields["email"] = email }
        if let phone { fields["phone"] = phone }
        return command(type: "visitor:update", fields: fields)
    }

    public static func storageSet(key: String, value: String) -> String {
        command(type: "storage:set", fields: ["key": key, "value": value])
    }
}

File: android/wiretalk-android/src/main/java/com/wiretalk/sdk/WiretalkBridge.kt
package com.wiretalk.sdk

import org.json.JSONObject

data class WiretalkBridgeEvent(
    val type: String,
    val raw: JSONObject,
) {
    val visitorUid: String? get() = raw.optString("visitor_uid").takeIf { it.isNotBlank() }
    val conversationId: Long? get() = raw.optLong("conversation_id").takeIf { raw.has("conversation_id") && !raw.isNull("conversation_id") }
    val unreadCount: Int get() = raw.optInt("count", 0)
    val messageId: Long? get() = raw.optLong("message_id").takeIf { raw.has("message_id") }
    val messageBody: String? get() = raw.optString("body").takeIf { it.isNotBlank() }
    val isOpen: Boolean get() = raw.optBoolean("open", false)
    val tab: String? get() = raw.optString("tab").takeIf { it.isNotBlank() }
}

fun interface WiretalkBridgeListener {
    fun onEvent(event: WiretalkBridgeEvent)
}

object WiretalkBridge {
    fun parse(payload: String): WiretalkBridgeEvent? {
        return try {
            val json = JSONObject(payload)
            val type = json.optString("type")
            if (type.isBlank()) null else WiretalkBridgeEvent(type, json)
        } catch (_: Exception) {
            null
        }
    }

    fun buildCommand(type: String, fields: Map<String, Any?> = emptyMap()): String {
        val json = JSONObject()
        json.put("wiretalk", true)
        json.put("type", type)
        fields.forEach { (key, value) ->
            when (value) {
                null -> Unit
                is Number -> json.put(key, value)
                is Boolean -> json.put(key, value)
                else -> json.put(key, value.toString())
            }
        }
        return """
            window.dispatchEvent(new MessageEvent('message', { data: ${json} }));
        """.trimIndent()
    }

    fun open(tab: String = "chat") = buildCommand("widget:open", mapOf("tab" to tab))

    fun close() = buildCommand("widget:close")

    fun toggle() = buildCommand("widget:toggle")

    fun updateVisitor(name: String? = null, email: String? = null, phone: String? = null): String {
        val fields = mutableMapOf<String, Any?>()
        name?.let { fields["name"] = it }
        email?.let { fields["email"] = it }
        phone?.let { fields["phone"] = it }
        return buildCommand("visitor:update", fields)
    }

    fun storageSet(key: String, value: String) = buildCommand("storage:set", mapOf("key" to key, "value" to value))

    fun storageRemove(key: String) = buildCommand("storage:remove", mapOf("key" to key))
}
